presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java

```diff
@@ -148,6 +222,319 @@ private static void generateProcessMethod(ClassDefinition classDefinition, RowEx
                 .append(position.ret());
     }
 
+    private static void generateProcessColumnarMethod(
+            ClassDefinition classDefinition,
+            CallSiteBinder callSiteBinder,
+            RowExpression filter,
+            List<RowExpression> projections,
+            List<MethodDefinition> projectionMethods)
+    {
+        Parameter session = arg("session", ConnectorSession.class);
+        Parameter page = arg("page", Page.class);
+        Parameter types = arg("types", List.class);
+        MethodDefinition method = classDefinition.declareMethod(a(PUBLIC), "processColumnar", type(Page.class), session, page, types);
+
+        Scope scope = method.getScope();
+        ByteCodeBlock body = method.getBody();
+        Variable thisVariable = method.getThis();
+
+        // extract blocks
+        List<Integer> allInputChannels = getInputChannels(concat(projections, ImmutableList.of(filter)));
+        ImmutableMap.Builder<Integer, Variable> channelBlockBuilder = ImmutableMap.builder();
+        for (int channel : allInputChannels) {
+            Variable blockVariable = scope.declareVariable(Block.class, "block_" + channel);
+            body.append(blockVariable.set(page.invoke("getBlock", Block.class, constantInt(channel))));
+            channelBlockBuilder.put(channel, blockVariable);
+        }
+        Map<Integer, Variable> channelBlock = channelBlockBuilder.build();
+        Map<RowExpression, List<Variable>> expressionInputBlocks = getExpressionInputBlocks(projections, filter, channelBlock);
+
+        Variable selectedPositions = scope.declareVariable(int[].class, "selectedPositions");
+        body.append(selectedPositions.set(thisVariable.invoke("filterPage", int[].class, session, page)));
+
+        Variable cardinality = scope.declareVariable(int.class, "cardinality");
+        body.append(cardinality.set(selectedPositions.length()));
+
+        Variable positionCount = scope.declareVariable(int.class, "positionCount");
+        body.append(positionCount.set(page.invoke("getPositionCount", int.class)));
+
+        body.comment("if no rows selected return null")
+                .append(new IfStatement()
+                        .condition(equal(cardinality, constantInt(0)))
+                        .ifTrue(constantNull(Page.class).ret()));
+
+        if (projectionMethods.isEmpty()) {
+            // if no projections, return new page with selected rows
+            body.append(newInstance(Page.class, cardinality, newArray(type(Block[].class), 0)).ret());
+            return;
+        }
+
+        Variable pageBuilder = scope.declareVariable(PageBuilder.class, "pageBuilder");
+        body.append(pageBuilder.set(newInstance(PageBuilder.class, types)));
```

Comment:
We should add a constructor to `PageBuilder` where we can set the `expectedEntries` for each block builder.  You can add this in another PR.